### PR TITLE
Make couch_log configurable

### PR DESCRIPTION
**Pull request by kxepal**
This allows to set logging backend and level via config API making
logging configuration a bit closer as it was in CouchDB 1.x.

### CODE REVIEW

**Comment by kxepal** [comment]
cc @eiri @jaydoane

**Comment by mikewallace1979** [comment]
+1

**Comment by jaydoane** [comment]
I tested this patch locally, and it seems to work fine. Specifically, the [log] level = none in rel/files/eunit.ini suppresses logging during test runs.
+1

**Comment by kxepal** [comment]
Good to hear! I'll fix few more issues here (need a better level name cast and fix couch_log_app hack), test a bit more and then will merge. Stay tuned! (:

**Comment by kxepal** [comment]
Ok, should be better now.

File: src/couch_log.app.src [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+% distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+% WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+% License for the specific language governing permissions and limitations under
+% the License.
+
+{application, couch_log, [
+    {description, "CouchDB Log API"},
+    {vsn, git},
+    {modules, [
+        couch_log,
```
Comment (rnewson) [inline src/couch_log.app.src]: empty this out, rebar will populate it for us automatically, it can, at best, be duplication.

**Comment by rnewson** [comment]
This does more than make the backend runtime-configurable (I'm not sure why that would be useful tbh), it also makes the top-level very prescriptive about log level filtering. I don't really like it.

**Comment by kxepal** [comment]
@rnewson 

It does exactly runtime configurable logging and only. However, you cannot just add few calls to `config:get` in couch_log because:
1. Somehow need to notify backends about changes in logging level. Lager has own configuration file where level is set (now it's error). So if you do "set backend stderr, set level warning, set backend lager", lager won't get log level update and you have to reset it.
2. Subscription on config changes need to be done somewhere on the start. Who and where should do that? The only reasonable candidate is couch_log itself, so that's why I have to introduce app and sup.
3. You need somehow tolerate user input.

So it indeed can be limited by functional, but it will also have holes to break the things.

While backend change in runtime is indeed not much useful, log level changes is quite handy feature. Asking people to use remsh to change log level is not very friendly way to go.

Filtering log messages on such top level serves two purposes:
1. Don't lie in the stats. Current behavior is that we trigger the counter while there nothing actually get logged because of level restrictions. That is very strange to record activity that didn't happened and confusing.
2. Use short circuit if we can. Since we make CouchDB config as authoritative source of information about logging level, we can avoid call any backend handler if we know that this give no effect.

In anyway, we should limited our ways to configure things and logging. Use OS environment variables, application environment, special files is not a good sign. After this, we have INI file only + special file for lager until we figure something better. 

Any alternative solutions are possible here?

File: src/couch_log.app.src (hunk 1 above)
Comment (kxepal) [inline src/couch_log.app.src]: Ok

**Comment by kxepal** [comment]
Ok, since no alternatives suggested, let's make it so and improve later.

**Pull request merged**